src/pydocstyle/parser.py

```diff
@@ -117,7 +118,17 @@ def is_public(self):
 
         This helps determine if it requires a docstring.
         """
-        return not self.name.startswith('_') or self.name.startswith('__')
+        module_name = Path(self.name).name
+        return (
+            not self._is_inside_private_package() and
+            (not module_name.startswith('_') or module_name.startswith('__'))
+        )
+
+    def _is_inside_private_package(self):
+        """Return True if the module is inside a private package."""
+        path = Path(self.name)
+        package_names = path.parts[:-1]
+        return any([package.startswith('_') for package in package_names])
```

Comment:
How does this handle any directories starting with `__`?

Reply:
> How does this handle any directories starting with `__`?

It considers them **private**, and I added a unit test for this in 84fa77f07ecd2d7373a793d1130ec61a75a14bac.

Should they be considered **public** instead like modules starting with double-underscores?

Reply:
Modules with double underscore should be considered **public** (e.g., `__main__.py`, `__init__.py`, etc.)

Reply:
Thanks for the clarification @Nurdok.

Considering *special* "dunder" modules like `__init__.py` **public** makes sense.

Do you consider modules like `__my_module.py`, *without* the *trailing* double-underscores,  **public** or **private**?

Reply:
I would keep any module that begins and ends with a double underscore public and the rest should be considered private. So something like `__main__.py` should be public but something like `__my_module.py` would hopefully be private.

Reply:
> So something like `__main__.py` should be **public** but something like `__my_module.py` would hopefully be **private**.

Updated in https://github.com/PyCQA/pydocstyle/pull/494/commits/ef7f5a1945e2a20f42e5913796892abc478656f9.